### ISSUE TITLE
WPE 2.38 - Set URL fails

### ISSUE THREAD

**Issue by amol-virnodkar-infosys** [Metrological/WebKitForWayland#1334]
The Set URL fails on RPI4 Model B device

**Test steps**

1. Launch the UIController using a browser
2. Enable WebKitBrowser plugin using the Controller
3. Select WebKitBrowser plugin in the UIController
4. Set the Custom URL to http://mvt.onemw.net/test-materials/sandbox/navigation_test.html 
5. Get the current Browser URL 
```
root@raspberrypi4-64-rdk-android-mc:~# curl -d '{"method": "WebKitBrowser.1.url"}'  http://localhost:9998/jsonrpc
\{"jsonrpc":"2.0","result":"http:\/\/mvt.onemw.net\/test-materials\/sandbox\/navigation_test.html#676"}
root@raspberrypi4-64-rdk-android-mc:~# 
```
6. Select WebKitBrowser plugin in the UIController, set the Custom URL to https://widgets.metrological.com/lightning/liberty/2e3c4fc22f0d35e3eb7fdb47eb7d4658#app:com.metrological.app.CNN 
7. Get the current Browser URL
```
root@raspberrypi4-64-rdk-android-mc:~# curl -d '{"method": "WebKitBrowser.1.url"}'  http://localhost:9998/jsonrpc
{"jsonrpc":"2.0","result":"http:\/\/mvt.onemw.net\/test-materials\/sandbox\/navigation_test.html#7372"}
root@raspberrypi4-64-rdk-android-mc:~# 
```

Ideally after step 6, the current Browser URL should be set to https://widgets.metrological.com/lightning/liberty/2e3c4fc22f0d35e3eb7fdb47eb7d4658#app:com.metrological.app.CNN but the curl command in step 7 is returning the URL as http://mvt.onemw.net/test-materials/sandbox/navigation_test.html#7372

**Note**
RDK6 comes with WPE 2.38 and we have used RDK6 image on RPI4 device. This issue is observed using the RDK6 image.

**Comment by amol-virnodkar-infosys** [comment]
Note: The source code of http://mvt.onemw.net/test-materials/sandbox/navigation_test.html consists of changeLocation() function which is called every 10ms.

```
<html>

<head>
    <script>
        let i = 0;
        function changeLocation() {
            i += 1
            window.location.href = "#" + i;
            setTimeout(changeLocation, 10)
        }
        changeLocation()
    </script>
</head>

<body>
    Navigation test
</body>

</html>
```

**Comment by pgorszkowski-igalia** [comment]
I am able to reproduce the problem on RPi3 with wpe-2.38 and wpe-2.42 with WPEFramework. I use a local version of the test site with the same content as it is in http://mvt.onemw.net/test-materials/sandbox/navigation_test.html. Steps to reproduce:

1. Launch the UIController using a browser
2. Enable WebKitBrowser plugin using the Controller
3. Select WebKitBrowser plugin in the UIController
4. Set the Custom URL to http://localhost/navigation_test.html
5. Set the Custom URL to https://www.igalia.com

After step 4 there is no possibility to switch to other site like https://www.igalia.com in step 5. It is not possible also with webinspector and `window.location.href="https://www.igalia.com"`

I am investigating the reason of that and possible solution.

**Comment by magomez** [comment]
Does this reproduce if the env var WPE_KEEP_NAVIGATION_ON_FRAGMENT_LOAD=1 is defined?. I think commit https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/525d2d60d8387773a023f7ae67c74c3b06430457 was added precisely to fix this problem.

**Comment by pgorszkowski-igalia** [comment]
It fixes partially the problem.  But even then I can reproduce the problem with lower probability - still checking this.

**Comment by amol-virnodkar-infosys** [comment]
I tried setting the environment variable WPE_KEEP_NAVIGATION_ON_FRAGMENT_LOAD=1 and retested. With this change, there is some improvement seen but the issue was reproducible 5/10 times.

**Comment by pgorszkowski-igalia** [comment]
The issue I reported also in upstream and I am working on the fix: https://bugs.webkit.org/show_bug.cgi?id=274745

**Comment by pgorszkowski-igalia** [comment]
Hi @amol-virnodkar-infosys , can you check my patch from: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1354 and verify it works for you and that there are no regressions? thanks

Please remember to set WPE_KEEP_NAVIGATION_ON_FRAGMENT_LOAD=1.

**Comment by amol-virnodkar-infosys** [comment]
Hi @pgorszkowski-igalia,

We tested this patch: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1354 with env. variable WPE_KEEP_NAVIGATION_ON_FRAGMENT_LOAD=1 and this issue is not reproducible. There are no regressions seen due to this patch as well. Thanks.

**Comment by pgorszkowski-igalia** [comment]
Hi @amol-virnodkar-infosys, thanks for the feedback. In that case we will merge PR https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1354.

**Comment by pgorszkowski-igalia** [comment]
The fix is merged so we can close the ticket.

**Comment by amol-virnodkar-infosys** [comment]
While further testing, if we set the url as "https://widgets.metrological.com/lightning/liberty/2e3c4fc22f0d35e3eb7fdb47eb7d4658#app:com.gametree.tv.Tetris" using the curl command, we observed black screen while opening this application on WPE 2.38 browser. 
Could you please let us know if we need to create a new ticket for this issue.

**Comment by pgorszkowski-igalia** [comment]
@amol-virnodkar-infosys : is the problem with Tetris application reproducible without my fix or without WPE_KEEP_NAVIGATION_ON_FRAGMENT_LOAD?

**Comment by pgorszkowski-igalia** [comment]
I have checked it. It is a regression after https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1354.

**Comment by amol-virnodkar-infosys** [comment]
@pgorszkowski-igalia Yes, it is a regression caused by https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1354. We observed the same behaviour.

**Comment by pgorszkowski-igalia** [comment]
@amol-virnodkar-infosys : can you check and verify https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/620bf81ccc14b76ff61c6c40a389a5b754fced1b? Remember to use: WPE_KEEP_NAVIGATION_ON_FRAGMENT_LOAD

**Comment by amol-virnodkar-infosys** [comment]
With the patch provided in https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/620bf81ccc14b76ff61c6c40a389a5b754fced1b and used WPE_KEEP_NAVIGATION_ON_FRAGMENT_LOAD, curl command with Tetris URL as "https://widgets.metrological.com/lightning/liberty/2e3c4fc22f0d35e3eb7fdb47eb7d4658#app:com.gametree.tv.Tetris" is working. 

We observed below two failures

A. Failure to load Sudoku App
1. Set the Custom URL to Sudoku App(https://widgets.metrological.com/lightning/liberty/2e3c4fc22f0d35e3eb7fdb47eb7d4658#app:com.metrological.app.Sudoku) using curl & Sudoku application fails to load. (This was observed on LG box)

B. The following sequence fails to load the Tetris application
1. Set the Custom URL to http://mvt.onemw.net/test-materials/sandbox/navigation_test.html
2. Set the Custom URL to CNN App
3. Set the Custom URL to Tetris App(https://widgets.metrological.com/lightning/liberty/2e3c4fc22f0d35e3eb7fdb47eb7d4658#app:com.gametree.tv.Tetris)

Tetris app fails to load on screen

**Comment by pgorszkowski-igalia** [comment]
I think that https://widgets.metrological.com/lightning/liberty/2e3c4fc22f0d35e3eb7fdb47eb7d4658#app:com.metrological.app.Sudoku is problematic even without this change, can you confirm that?
My test was (the build without my changes provided here):
1. Set the Custom URL to https://www.igalia.com
2. Set the custom URL to CNN App
3. Set the Custom URL to Tetris App(https://widgets.metrological.com/lightning/liberty/2e3c4fc22f0d35e3eb7fdb47eb7d4658#app:com.gametree.tv.Tetris)

Tetris app fails to load on screen and WPE is not responding so it is killed (WPEFramework sends SIGFPE signal to kill it because of lack of responsiveness)

**Comment by amol-virnodkar-infosys** [comment]
I used RPI4 image without the changes.
```
root@raspberrypi4-64-rdk-android-mc:~# cat /version.txt 
imagename:rdk-generic-mediaclient-wpe-image_FBT_TDK_rdk6-main_20240316134950
BRANCH=rdk6-main
YOCTO_VERSION=dunfell
VERSION=rdk6-main.03.16.24
SPIN=0
BUILD_TIME="2024-03-16 13:49:50"
WPE_WEBKIT_VERSION=2.38.2+gitAUTOINC+0ac2e6f2b6-r12
Generated on Sat Mar 16  13:49:50 UTC 2024
JENKINS_JOB=0
JENKINS_BUILD_NUMBER=0
root@raspberrypi4-64-rdk-android-mc:~# 

```
Please find below observations on RPI4.
1. _**Sudoku**_
    Setting the custom URL from the Controller is working. But using the curl command setting of Sudoku URL is not working.
On executing the below command we can see URL is set to "about:blank"
```
root@raspberrypi4-64-rdk-android-mc:~# curl -d '{"method": "WebKitBrowser.1.url", "params": "https://widgets.metrological.com/lightning/liberty/2e3c4fc22f0d35e3eb7fdb47eb7d4658#app:com.metrological.app.Sudoku"}' http://127.0.0.1:9998/json
root@raspberrypi4-64-rdk-android-mc:~# 
root@raspberrypi4-64-rdk-android-mc:~# 
root@raspberrypi4-64-rdk-android-mc:~# 
root@raspberrypi4-64-rdk-android-mc:~# curl -d '{"method": "WebKitBrowser.1.url"}'  http://localhost:9998/jsonrpc
{"jsonrpc":"2.0","result":"about:blank"}root@raspberrypi4-64-rdk-android-mc:~# 
```

2. I executed below steps using Controller UI (Similar behaviour seen on RPI4 & LG box)
a. Set the Custom URL to https://www.igalia.com/
    Igalia website is launched in the browser.
b. Set the custom URL to CNN App
    CNN App is launched in the browser.
c. Set the Custom URL to Tetris App(https://widgets.metrological.com/lightning/liberty/2e3c4fc22f0d35e3eb7fdb47eb7d4658#app:com.gametree.tv.Tetris)
    Tetris App is not loaded on the screen but using curl command we can see that URL is set to Tetris application. No crash was observed due to SIGFPE.

**Comment by pgorszkowski-igalia** [comment]
The problem with Sudoku application is not related to this issue and my fix. It should be reported separately.

**Comment by amol-virnodkar-infosys** [comment]
@pgorszkowski-igalia Sure, I will report the issue with Sudoku application separately. Thanks

**Comment by pgorszkowski-igalia** [comment]
@amol-virnodkar-infosys : can we merge https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/620bf81ccc14b76ff61c6c40a389a5b754fced1b as better solution for this case (assuming that the case with Sudoku app is something different)?

**Comment by amol-virnodkar-infosys** [comment]
@pgorszkowski-igalia As suggested by you since https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/620bf81ccc14b76ff61c6c40a389a5b754fced1b is a better solution, we can merge this change. We will take the case with Sudoku app in a separate ticket.

**Comment by pgorszkowski-igalia** [comment]
@amol-virnodkar-infosys : can you apply https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1374 and check there are no regressions with it?

**Comment by amol-virnodkar-infosys** [comment]
Tested the WPE 2.38 browser with the changes - https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/620bf81ccc14b76ff61c6c40a389a5b754fced1b and https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1374, the issue is not reproducible. The other tests are also working as expected. Thanks.

**Comment by pgorszkowski-igalia** [comment]
https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1374 and https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1386 are merged in wpe-2.38 so I close this issue.
